### PR TITLE
Allow escaping commas in crypttab

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/basic/strv.c [hunk 1]
```diff
@@ -378,11 +378,17 @@ char *strv_join_prefix(char * const *l, const char *separator, const char *prefi
         k = strlen(separator);
         m = strlen_ptr(prefix);
 
+        if (unescape_separators) /* If there separator is multi-char, we won't know how to escape it. */
+                assert(k == 1);
```
Comment (poettering) [inline src/basic/strv.c:382]: given that asserts in theory can be optimized away, this one is a bit weird. maybe:

 ```c
assert(!unescape_separators || k == 1);
```

File: src/basic/strv.c [hunk 2]
```diff
@@ -397,7 +403,16 @@ char *strv_join_prefix(char * const *l, const char *separator, const char *prefi
...
+                        for (size_t i = 0; (*s)[i]; i++) {
+                                if ((*s)[i] == separator[0])
+                                        *(e++) = '\\';
+                                *(e++) = (*s)[i];
+                        }
+                else
+                        e = stpcpy(e, *s);
         }
 
         *e = 0;
```
Comment (poettering) [inline src/basic/strv.c:418]: isn't this a bit ad-hoc? i figure we should escape spaces/tabs/newlines and so on too when used in cryptsetup cmdlines?

i.e. shouldn't this use cescape() or xescape() or some form of it? i.e. first each elements is escaped with these functions and then joined afterwards?

File: src/shared/fstab-util.c [hunk 3]
```diff
@@ -95,7 +95,19 @@ int fstab_filter_options(const char *opts, const char *names,
...
+                         * the only valid escape sequence, so we can do a very simple test here. */
+                        for (;;) {
+                                size_t n = strcspn(end, ",");
+
+                                end += n;
+                                if (n > 0 && end[-1] == '\\')
+                                        end++;
+                                else
+                                        break;
+                        }
```
Comment (poettering) [inline src/shared/fstab-util.c:110]: this feels pretty adhoc too. at the very least this should be a function of its own

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/basic/strv.c (hunk 1 above)
Comment (keszybz) [reply src/basic/strv.c:382]: So... I know that this is the canonical form, but I find the version with separate `if` easier. The compiler doesn't care (and it should be smart enough to figure out that the conditional body is empty when compiled without asserts).

File: src/basic/strv.c (hunk 2 above)
Comment (keszybz) [reply src/basic/strv.c:418]: I think commas as the only character we will realistically need to escape. The need to escape commas is caused by the mismatch between the kernel names which are passed as commandline arguments like `--cipher=… --foobar` where commas are unambiguous and our format where we squish multiple such arguments into a single word in crypttab: `cipher=…,foobar`. OTOH, whitespace and non-printable characters would be very awkward on the commandline and it's very unlikely that such values will be added. Thus, I think just allowing commas to be escaped is enough to solve the problem. It *is* ad-hoc, but I don't think we need to be a fully generic syntax here: being simple and unambiguous is important too.

File: src/shared/fstab-util.c (hunk 3 above)
Comment (keszybz) [reply src/shared/fstab-util.c:110]: The function is not allowed here, so a helper which allocates cannot be used. OTOH, a helper that just returns the length would replace 3–4 lines, which doesn't seem worth the trouble.

**Pull request merged**